### ISSUE TITLE
file:/// protocol when using Cesium off the web

### ISSUE THREAD

**Issue by TomPed** [AnalyticalGraphicsInc/cesium#3935]
This issue has to do with Cesium being used off the web, i.e., [Electron](http://electron.atom.io/) and [Cordova](https://cordova.apache.org/) trying to load the Natural Earth II imagery.

When you try loading the offline imagery you get this error: `net::ERR_FILE_NOT_FOUND`. One of the files not found are: `file://users/tpedbereznak/Cesium-Editor/node_modules/cesium/Build/CesiumUnminified/Assets/Textures/NaturalEarthII/tilemapresource.xml`. If you change `file://` to `file:///` then the file is found. Knowing this I made a hacky fix.

In [`loadXML.js`](https://github.com/AnalyticalGraphicsInc/cesium/blob/03191f3c7558d8ab6b3fe6fb002261501a2d9eb6/Source/Core/loadXML.js) I change line [line 37](https://github.com/AnalyticalGraphicsInc/cesium/blob/03191f3c7558d8ab6b3fe6fb002261501a2d9eb6/Source/Core/loadXML.js#L37): `url : url,` to `url : url.slice(0, 7) + "/" + url.slice(7),`.

In [`UrlTemplateImageryProvider.prototype.reinitialize`](https://github.com/AnalyticalGraphicsInc/cesium/blob/7fc81f744dca4bf2a9341a5837845ebdddf8a8e7/Source/Scene/UrlTemplateImageryProvider.js#L492) I changed line [line 504](https://github.com/AnalyticalGraphicsInc/cesium/blob/7fc81f744dca4bf2a9341a5837845ebdddf8a8e7/Source/Scene/UrlTemplateImageryProvider.js#L504): `that._url = properties.url;` to `that._url = properties.url.slice(0, 7)  + "/" + properties.url.slice(7);`

Now url's with `file://` with be changed to `file:///` and load the Natural Imagery II. How can this be fixed without this hacky solution?

**Comment by mramato** [comment]
What does the code you are using to load the imagery look like?  I suspect this is a problem at the app level.

**Comment by TomPed** [comment]
I just select Natural Earth II from the imagery picker.

**Comment by mramato** [comment]
Are you setting `CESIUM_BASE_URL` in your app?

**Comment by TomPed** [comment]
No, I'm not. Where should I be setting this?

**Comment by mramato** [comment]
I'm not sure that's the problem; I'm just trying to figure out what the issue could be.

**Comment by TomPed** [comment]
I just tried that `Cesium.buildModuleUrl.setBaseUrl('./node_modules/cesium/Build/CesiumUnminified');`. I think we can rule this out.

**Comment by mramato** [comment]
If you wanted to try setting it, I would use the absolute path with the 3 slashes.

**Comment by TomPed** [comment]
I tried: `Cesium.buildModuleUrl.setBaseUrl('file:///Users/tpedbereznak/Electron-Cesium-Starter-Kit/node_modules/cesium/Build/Cesium');`. Did not work either.

**Comment by hpinkos** [comment]
@TomPed is this still a problem or did you find a solution?

**Comment by TomPed** [comment]
Yes, this is still a problem, I have that *hacky* fix that I mentioned, but that's it.

**Comment by hpinkos** [comment]
Okay thanks, just checking =)

**Comment by twpayne** [comment]
I think the problem here that `file://` URLs can only use absolute paths. See [the Wikipedia article on the file URI scheme](https://en.wikipedia.org/wiki/File_URI_scheme) for more information, but the tl;dr is that `file://users/tpedbereznak/Cesium-Editor/node_modules/`... URI does not work because it interprets `users` as a hostname and `/tpedbereznak/Cesium-Editor/node_modules/`... as an absolute path on that host, which obviously fails (neither the host nor the path exists).

By hacking in an extra `/`, you change this to an absolute path *that only works on your machine*.

Instead, I suggest that you use a relative URL, without specifying the protocol (i.e. without `file://`) and make sure that you're not calling `require.toUrl` or similar function. You will need to find the correct relative URL.

I don't think this is a bug in Cesium.

**Comment by Enoooch** [comment]
Using three slashes `file:///` works on Windows, two slashes `file://` works on Mac.

Tested in cesium 1.87 `new Cesium.Cesium3DTileset()`.
